qiskit/circuit/quantumcircuitdata.py

```diff
@@ -13,21 +13,95 @@
 """A wrapper class for the purposes of validating modifications to
 QuantumCircuit.data while maintaining the interface of a python list."""
 
+from abc import abstractmethod
 from collections.abc import MutableSequence
 
 from qiskit.circuit.exceptions import CircuitError
 from qiskit.circuit.instruction import Instruction
 
 
-class QuantumCircuitData(MutableSequence):
+class ValidatedMutableSequence(MutableSequence):
+    """A wrapper for validating and acting upon modifications to an
+    underlying List.
+
+    Intended to be used when an underlying list structure, owned by
+    the QuantumCircuit (such as QuantumCircuit._data), is fetched by a
+    user (through e.g. the QuantumCircuit.data property) and may be
+    subsequently updated, such that updates through the returned
+    ValidatedMutableSequence are checked to ensure consistency with
+    the QuantumCircuit instance.
+    """
+
+    def __init__(self, list_):
+        self._list = list_
+
+    def __getitem__(self, i):
+        return self._list[i]
+
+    @abstractmethod
+    def __delitem__(self, i):
+        pass
+
+    @abstractmethod
+    def __setitem__(self, index, value):
+        pass
+
+    @abstractmethod
+    def insert(self, index, value):
+        pass
```

Comment:
That's true. I left the methods that mutated `self._list` as abstract because I couldn't see an obvious implementation that worked cleanly (across the few use cases we currently support) with validation, so this is left to the subclass.